Extending built-in classes (#237)
javascript-tutorial/it.javascript.info: 1-js/09-classes/05-extend-natives/article.md

alert(filteredArr.isEmpty()); // false
```

> Notiamo una cosa interessante. I metodi built-in come `filter`, `map` e così via -- ritornano nuovi oggetti del tipo ereditato, cioè `PowerArray`. La loro implementazione interna utilizzata il `costruttore` per farlo.

Reviewer: ```suggestion
Notiamo una cosa interessante. I metodi built-in come `filter`, `map` e così via, ritornano nuovi oggetti del tipo ereditato, cioè `PowerArray`. La loro implementazione interna utilizzata la proprietà oggetto `constructor` per farlo.
```

> Notiamo una cosa interessante. I metodi built-in come like `filter`, `map` e cosi via -- ritornano nuovi esattamente del nuovo tipo ereditato, cioè `PowerArray`. La loro implementazione interna utilizzata il `costruttore` per farlo.

Reviewer 2: ```suggestion
Notiamo una cosa interessante. I metodi built-in come `filter`, `map` e così via -- ritornano nuovi oggetti del tipo ereditato, cioè `PowerArray`. La loro implementazione interna utilizzata il `costruttore` per farlo.
```

Nell'esempio sopra,
```js
arr.constructor === PowerArray
```

> Quando invochiamo `arr.filter()`, questo creerà internamente il nuovo array contenente i risultati utilizzando `arr.constructor`, non l'`Array` standard. Questo è molto utile, poiché possiamo continuare ad utilizzare i metodi di `PowerArray` anche dopo aver ottenuto il risultato.

Reviewer: ```suggestion
Quando invochiamo `arr.filter()`, questo creerà internamente il nuovo array contenente i risultati utilizzando `arr.constructor`, non l'oggetto `Array` standard. Questo è molto utile, poiché possiamo continuare ad utilizzare i metodi di `PowerArray` anche dopo aver ottenuto il risultato.
```

> Ancora meglio, possiamo personalizzarne il comportamento.

Reviewer: ```suggestion
Inoltre, possiamo personalizzarne il comportamento.
```

> Possiamo aggiungre uno speciale getter statico `Symbol.species` alla classe. Questo dovrebbe ritornare il costruttore utilizzerà internamente per creare le nuove entità in  `map`, `filter` e gli altri metodi.

Reviewer 2: ```suggestion
Possiamo aggiungere uno speciale getter statico `Symbol.species` alla classe. Questo dovrebbe ritornare il costruttore che Javascript utilizzerà internamente per creare le nuove entità in  `map`, `filter` e gli altri metodi.
```

> Possiamo aggiungere uno speciale getter statico `Symbol.species` alla classe. Questo dovrebbe ritornare il costruttore che Javascript utilizzerà internamente per creare le nuove entità in  `map`, `filter` e gli altri metodi.

Reviewer: ```suggestion
Possiamo aggiungere uno speciale getter statico `Symbol.species` alla classe. Questo dovrebbe ritornare il costruttore che Javascript utilizzerà internamente per creare le nuove entità in  `map`, `filter` e così via.
```

> Se invece, la nostra intenzione è quella di restituire degli array standard tramite l'utilizzo dei metodi come `map` o `filter`, possiamo ritornare `Array` in `Symbol.species`, come nell'esempio:

Reviewer 2: ```suggestion
Se, invece, la nostra intenzione è quella di restituire degli array standard tramite l'utilizzo dei metodi come `map` o `filter`, possiamo ritornare `Array` in `Symbol.species`, come nell'esempio:
```
Una bella virgoletta in piu' :P

> Se, invece, la nostra intenzione è quella di restituire degli array standard tramite l'utilizzo dei metodi come `map` o `filter`, possiamo ritornare `Array` in `Symbol.species`, come nell'esempio:

Reviewer: ```suggestion
Se, ad esempio, volessimo che metodi come `map` o `filter` restituiscano un array standard, possiamo ritornare `Array` in `Symbol.species`, come nell'esempio:
```

let arr = new PowerArray(1, 2, 5, 10, 50);
alert(arr.isEmpty()); // false

// filter crea un nuovo array utilizzando arr.constructor[Symbol.species] come costruttore
let filteredArr = arr.filter(item => item >= 10);

*!*
// filteredArr non è di tipo PowerArray, ma è un Array standard
*/!*
alert(filteredArr.isEmpty()); // Error: filteredArr.isEmpty is not a function
```

Come potete osservare, ora `.filter` restituisce un `Array`. Quindi l'estensione delle funzionalità non sarà più disponibile.

```smart header="Le altre collezioni funzionano in maniera simile"
Le altre collezioni, come `Map` e `Set`, funzionano in maniera molto simile. Anche queste utilizzano `Symbol.species`.
```

## Con gli oggetti built-in non si ereditano le proprietà statiche

Gli oggetti built-in possiedono i loro metodi statici, ad esempio `Object.keys`, `Array.isArray` etc.

Come già sappiamo, le classi integrate si estendono a vicenda. Ad esempio, `Array` estende `Object`.

Normalmente, quando una classe ne estende un'altra, sia i metodi statici che quelli non-statici vengono ereditati. Questo è stato ampiamente spiegato nell'articolo [](info:static-properties-methods#statics-and-inheritance).

> Ma le classi built-in fanno eccezione. Queste infatti, non ereditano i membri statici a vicenda.

Reviewer 2: ```suggestion
Ma le classi built-in fanno eccezione. Queste, infatti, non ereditano i membri statici a vicenda.
```
altra virgoletta :P

> Ma le classi built-in fanno eccezione. Queste, infatti, non ereditano i membri statici a vicenda.

Reviewer: ```suggestion
Ma le classi built-in fanno eccezione. Queste, infatti, non ereditano i membri statici a le une dalle altre.
```
